pkg/cloudprovider/aws/aws.go

```diff
@@ -489,3 +808,338 @@ func (self *AWSCloud) GetZone() (cloudprovider.Zone, error) {
 		Region:        self.region.Name,
 	}, nil
 }
+
+// Gets the current load balancer state
+func (self *AWSCloud) describeLoadBalancer(region, name string) (*elb.LoadBalancer, error) {
+	loadBalancers, err := self.ec2.DescribeLoadBalancers(region, name)
+	if err != nil {
+		return nil, err
+	}
+
+	var ret *elb.LoadBalancer
+	for _, loadBalancer := range loadBalancers {
+		if ret != nil {
+			glog.Errorf("Found multiple load balancers with name: %s", name)
+		}
+		ret = &loadBalancer
+	}
+	return ret, nil
+}
+
+// TCPLoadBalancerExists implements TCPLoadBalancer.TCPLoadBalancerExists.
+func (self *AWSCloud) TCPLoadBalancerExists(name, region string) (bool, error) {
+	lb, err := self.describeLoadBalancer(name, region)
+	if err != nil {
+		return false, err
+	}
+
+	if lb != nil {
+		return true, nil
+	}
+	return false, nil
+}
+
+// Find the kubernetes vpc
+func (self *AWSCloud) findVpc() (*ec2.VPC, error) {
+	name := "kubernetes-vpc"
+	vpcs, err := self.ec2.ListVpcs(name)
+	if err != nil {
+		return nil, err
+	}
+	if len(vpcs) == 0 {
+		return nil, nil
+	}
+	if len(vpcs) == 1 {
+		return &vpcs[0], nil
+	}
+	return nil, fmt.Errorf("Found multiple matching VPCs for name: %s", name)
+}
+
+func mapToInstanceIds(instances []*ec2.Instance) []string {
+	ids := make([]string, 0, len(instances))
+	for _, instance := range instances {
+		ids = append(ids, instance.InstanceId)
+	}
+	return ids
+}
+
+func (self *AWSCloud) ensureSecurityGroupIngess(securityGroupId string, sourceIp string, protocol string, fromPort, toPort int) (bool, error) {
+	groups, err := self.ec2.DescribeSecurityGroups([]string{securityGroupId}, "", "")
+	if err != nil {
+		glog.Warning("error retrieving security group", err)
+		return false, err
+	}
+
+	if len(groups) == 0 {
+		return false, fmt.Errorf("security group not found")
+	}
+
+	if len(groups) != 1 {
+		return false, fmt.Errorf("multiple security groups found with same id")
+	}
+
+	group := groups[0]
+
+	for _, permission := range group.IPPerms {
```

Comment:
I added some comments to try to clarify.

We're adding permissions to an existing security group.  It is an error if the security group itself does not already exist.  It is also an error if we match multiple security groups, because this should be impossible.